### PR TITLE
Paged file browser

### LINKED ISSUE #14: Decrease RAM usage of Import Sample Dialog
Currently its not possible to open a directory containing more than about 25 samples because of the large amounts of ram used by the Import Sample dialog that came from LGPT.

This occurs for several reasons:
1. `I_Dir` is just a subclass of `T_SimpleList<Path>` so that everyfile in a directory is actually stored in the list with its full path not just the filename, so the containing directories path is stored in every files `Path` object
1. The whole contents of a directory are read into the `I_Dir` list when entering the directory, even though only 15 (`LIST_SIZE`) can be displayed at any one time by the UI, but this is necessary in order to be able to sort the file list for a dir
1. The `ImportSampleDialog` itself maintains another list `sampleList_` that it then copies (based on filter criteria, eg, if dir, if `*.wav`, the items from the list returned in the `I_Dir` object by `picoTrackerFileSystem`.

In my initial testing, it seems there is only a few (3-5?) KB of ram available for the ImportSampleDialog to use so the above code is very limiting in what size of dirs can be handled at the moment.

To try to improve this my plan is to create a new `PagedImportSampleDialog` class that uses paging, so that it only ever requests the 15 items it needs to display at any time. Then I need to create a new interface, `I_PagedDir` that supports being given a dir path and a s start index and it returns the 15 items from that start index. 

**But** the tricky thing here is: sorting. In order to be able to sort a directory, the new `I_PagedDir` would need to read and store every filename in a dir and that could just be itself run out of ram in large dirs: eg. for a dir with 300 files, average filename length of 20 chars gives: ~5.9KB and blows the ram budget. SDFat is capable of dealing with files just by a "directory index" `int` BUT then there is no way to alphabetically sort the contents :-(

So the 2 options I can see for the above are:

a. allow upto thousands of files per dir BUT they will not be sorted when displayed (except for subdirs which could at least be sorted to be at the start of the list)

b. limit the number of files based on filename lengths this could be anywhere from 20 to 100 but still allow for displaying them in alpha sorted order.

### PR DESCRIPTION
**Pull request by maks**
This signficantly reduces the ram usage when using the import dialog to browse and import samples from the sdcard.

It changes the Import Dialog when built for picotracker to use a new `picoTrackerPagedDir` class that implements a new interface instead of `I_Dir`.  This changes the file system operations to be done in "pages" of files and only stores the indexes of file and subdirectory entries of the currently open directory and only fetches the names of entries on demand in "pages" limited in size to what the dialog UI can display at one time (currently 15).

The `ImportSampleDialog` is also heavily rewritten to make use of the interface and use as little ram as possible.

This doesn't impose any hard limits on the number of files per directory, but it does set initial sizes of max of `96` subdirs and `354` files per subdir so its quite possible that exceeding these limits could cause extra memory allocation to overflow available ram at the moment.

Please note this is "stacked" on top of PR #26 so that PR should be merged before this one is.

Note: these changes do mean that apart from having subdirs listed before files, there is no other sorting of entries in a directory being applied for now.

Also I've kept all my commits in this PR but I would recommend doing a "squash merge" as all these commits don't really need to be included in the main branches history.

Fixes #14

### CODE REVIEW

**Comment by democloid** [comment]
For starters a couple of comments.
I should put this on some dev standards file or something. Could you use "clang-format" with default options on new files? I've been doing that in all new files and when I modify others, I regret not having done this on the base project before I started the fork.

How useful is the gui debug code enabled with SHOW_MEM_USAGE? Do you think this makes sense rather than debug logs? Not against per se, but maybe if it's something that we want to support for development we should create some generic define GUI_DEBUG? GUI_DEBUG_MESSAGES?

**Comment by maks** [comment]
Thanks for taking a look at this so quickly @democloid ! 👍🏻 

Ah thanks for the pointer, I didn't know about clang-format! I lean on dartfmt heavily when writing in Dart so great to know theres a tool for  c++ as well. I'll reformat all the new files with it. Its tempting for me to reformat existing file too as the mix of tabs and spaces drives me nuts, but the downside there is losing `git blame` to be able to quickly see if code was changed for picotracker or is original Lgpt code.

Oops, sorry I left `SHOW_MEM_USAGE` enable by mistake, will fix that 👍🏻 And double oops, forgot to disable uart debug as well.
But in general, yes I found it very useful to have on the screen during debugging as I had *alot* of debug logging enabled as I don't yet have uart for gdb setup so I was doing "print line debugging" and having mem usage on the display vs swapped by lots of other logs did help. So having a more general `#define` would probably be a good idea but for now I can't think of other UI I want to add so maybe just keep it for now until we need something more general?

**Comment by maks** [comment]
@democloid I've updated to remove the debug build settings and also reformatted my new files and the 2 main files I changed in the picotracker subdir with clang-format (default settings) but I kept the others as is to try to make it clearer what changes I made as part of this (except for `SelectProjectDialog` where I had already changed some of the whitespace formatting but didnt seem worth it to reformat the whole file with clang-format).

File: sources/Application/Views/ModalDialogs/SelectProjectDialog.cpp [hunk 1]
```diff
@@ -282,32 +281,27 @@ void SelectProjectDialog::setCurrentFolder(Path &path) {
...
-  {
-
+	if (dir) 
+	{
 		// Get all lgpt something
-
 		dir->GetContent("*");
-    dir->Sort();
-      
+		// dir->Sort();
```
Comment (democloid) [inline sources/Application/Views/ModalDialogs/SelectProjectDialog.cpp:288]: Should we keep this and define an empty sort method for the moment? (possibly with a log stating sorting is not implemented yet)

I'd leave this file untouched, most changes are unrelated

File: sources/System/FileSystem/FileSystem.h [hunk 2]
```diff
@@ -81,8 +93,8 @@ class I_File {
 public:
 	virtual ~I_File() {} ;
 	virtual int Read(void *ptr, int size, int nmemb)=0;
-  virtual int GetC() = 0;
-  virtual int Write(const void *ptr,int size, int nmemb)=0;
+	virtual int GetC() = 0;
```
Comment (democloid) [inline sources/System/FileSystem/FileSystem.h:96]: revert tabs

**Review by democloid — CHANGES_REQUESTED** [review]
Thanks for this PR! do you have a sense on how much total memory is being used in the worst case scenario?

I still think this should be made static mem, at the end of the day, even if this memory is used only in this part of the code, it's very hard to accurately keep track of how much heap is used by different components in order to effectively share that memory usage. The risks of weird bugs apearing is too high. We can leave that for later tho, as this will make the current situation better anyway and we can convert later.

File: sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp [hunk 3]
```diff
@@ -3,6 +3,123 @@
...
 
+#include <string>
+
+// Max filename is actually 256 per FAT std
+static const int MAX_FILENAME_LEN = 128;
+
+picoTrackerPagedDir::picoTrackerPagedDir(const char *path)
+    : path_{std::string(path)} {
+  fileIndexes_.reserve(354);
+  subdirIndexes_.reserve(96);
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:14]: Yes good point! It is very arbitrary, I just picked numbers that experimentally didnt cause any issues when I tested them. 
But these aren't actually limits per se, this just sets the initial capacity of the vectors, they will dynamically grow if items are added that exceed the current capacity, its just when that happens, that will likely exceed available ram, at least thats available at the moment, without further ram usage optimisations.

File: sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp [hunk 4]
```diff
@@ -3,6 +3,123 @@
...
+  }
+  if (!dir.isDir()) {
+    Trace::Error("Path:%s is not a directory", path_.c_str());
+    return;
+  }
+  static const int MAX_ITEMS = 15;
+  char current[MAX_FILENAME_LEN];
+  FsBaseFile file;
+
+  if (startOffset == 0 && (path_ != std::string("/samplelib"))) {
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:88]: oops, yes well spotted, I'll fix this up! 👍🏻

File: sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp [hunk 5]
```diff
@@ -3,6 +3,123 @@
 #include "Application/Utils/wildcard.h"
 #include "System/Console/Trace.h"
 
+#include <string>
+
+// Max filename is actually 256 per FAT std
+static const int MAX_FILENAME_LEN = 128;
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:9]: Yes, I use `MAX_FILENAME_LEN` for reading the filename for example when we need to actually open the file, though I'm no longer passing more than 24 chars out to the callers of this class, ie, SampleImportDialog as it can't actually display longer filenames as you mention.

File: sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp [hunk 6]
```diff
@@ -3,6 +3,123 @@
...
+
+  if (!dir.open(path_.c_str())) {
+    Trace::Error("getFileList failed open:%s", path_.c_str());
+    return;
+  }
+  if (!dir.isDir()) {
+    Trace::Error("Path:%s is not a directory", path_.c_str());
+    return;
+  }
+  static const int MAX_ITEMS = 15;
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:84]: I have a (bad?) habit of trying not to use the preprocessor, but yes it could be a define though I think it would basically come out as the same generated code I think, but I can change it to a define if you prefer that?

File: sources/Adapters/picoTracker/gui/picoTrackerGUIWindowImp.cpp
[the remark this answers is not in the document]
Comment (maks) [reply sources/Adapters/picoTracker/gui/picoTrackerGUIWindowImp.cpp:1]: Yes sorry that is due to stacking on previous PRs.

File: sources/Application/Model/Project.h [hunk 7]
```diff
@@ -17,7 +17,7 @@
 
 #define PROJECT_NUMBER "1.0"
 #define PROJECT_RELEASE "r"
-#define BUILD_COUNT "003"
+#define BUILD_COUNT "004"
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Model/Project.h:20]: oops yes! I think that got left in by mistake

File: sources/Application/Views/InstrumentView.cpp [hunk 8]
```diff
@@ -274,7 +275,8 @@ void InstrumentView::ProcessButtonMask(unsigned short mask,bool pressed) {
                 DoModal(mb) ;
               } else { ;
                 // Go to import sample
-                ImportSampleDialog *isd=new ImportSampleDialog(*this) ;
+                // ImportSampleDialog *isd=new ImportSampleDialog(*this) ;
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Views/InstrumentView.cpp:278]: oops yes well spotted! I meant to delete that before commiting.

File: sources/Application/Views/ModalDialogs/ImportSampleDialog.cpp
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Views/ModalDialogs/ImportSampleDialog.cpp:1]: yes thanks! will remove.

File: sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+#include "Application/Instruments/SamplePool.h"
+#include "pico/multicore.h"
+#include <memory>
+
+#define LIST_SIZE 15
+#define LIST_WIDTH 24
+
+const static char *sampleLib_ = "/samplelib";
+
+static const char *buttonText[3] = {"Listen", "Import", "Exit"};
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:12]: ah sorry, I didn't even notice that both were valid in c++. I'll make it consistent.

File: sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:1]: I kept it as I still want to have the orig implementation for now to use on desktop platforms as I still have hopes of using this code as the basis for a "modernised" lgpt for Linux desktop & game handhelds.

File: sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+void PagedImportSampleDialog::preview(Path &element) {
+  Player::GetInstance()->StartStreaming(element);
+}
+
+void PagedImportSampleDialog::import(Path &element) {
+
+  SamplePool *pool = SamplePool::GetInstance();
+
+#ifdef PICO_BUILD
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:143]: oops I should have removed that, thanks for spotting that 👍🏻

File: sources/Application/Views/ModalDialogs/PagedImportSampleDialog.h [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+protected:
+  void setCurrentFolder(Path *path);
+  void warpToNextSample(int dir);
+  void import(Path &element);
+  void preview(Path &element);
+
+private:
+  std::vector<FileListItem> fileList_{};
+  int currentSample_;
+  int topIndex_;
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Views/ModalDialogs/PagedImportSampleDialog.h:29]: yes good call! will fix it.

File: sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+        }
+        isDirty_ = true;
+        return;
+      }
+      auto fullPathStr = std::string(currentPath_.GetPath());
+      fullPathStr += "/";
+      fullPathStr += currentItem.name;
+      auto fullPath = Path{fullPathStr};
+
+      switch (selected_) {
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:207]: actually I have a follow up PR that gets rid of the buttons and then `selected_` will just go away

File: sources/Application/Views/ModalDialogs/SelectProjectDialog.cpp (hunk 1 above)
Comment (maks) [reply sources/Application/Views/ModalDialogs/SelectProjectDialog.cpp:288]: good point! will do.

Comment (maks) [reply sources/Application/Views/ModalDialogs/SelectProjectDialog.cpp:288]: actually this is still using the older I_Dir so it can still do the sorting

**Comment by maks** [comment]
@democloid I've hopefully addressed all the feedback, either with code changes or replies, but do let me know if you have followup comments 👍🏻 .

**Review by democloid — APPROVED** [review]
